cannon/mipsevm/state_mt.go

```diff
@@ -0,0 +1,293 @@
+	EXITCODE_MT_WITNESS_OFFSET                   = HEAP_MT_WITNESS_OFFSET + 4
+	EXITED_MT_WITNESS_OFFSET                     = EXITCODE_MT_WITNESS_OFFSET + 1
+	STEP_MT_WITNESS_OFFSET                       = EXITED_MT_WITNESS_OFFSET + 1
+	STEPS_SINCE_CONTEXT_SWITCH_MT_WITNESS_OFFSET = STEP_MT_WITNESS_OFFSET + 8
+	WAKEUP_MT_WITNESS_OFFSET                     = STEPS_SINCE_CONTEXT_SWITCH_MT_WITNESS_OFFSET + 8
+	TRAVERSE_RIGHT_MT_WITNESS_OFFSET             = WAKEUP_MT_WITNESS_OFFSET + 4
+	LEFT_THREADS_ROOT_MT_WITNESS_OFFSET          = TRAVERSE_RIGHT_MT_WITNESS_OFFSET + 1
+	RIGHT_THREADS_ROOT_MT_WITNESS_OFFSET         = LEFT_THREADS_ROOT_MT_WITNESS_OFFSET + 32
+	THREAD_ID_MT_WITNESS_OFFSET                  = RIGHT_THREADS_ROOT_MT_WITNESS_OFFSET + 32
+)
+
+type MTState struct {
+	Memory *Memory `json:"memory"`
+
+	PreimageKey    common.Hash `json:"preimageKey"`
+	PreimageOffset uint32      `json:"preimageOffset"` // note that the offset includes the 8-byte length prefix
+
+	Heap uint32 `json:"heap"` // to handle mmap growth
+
+	ExitCode uint8 `json:"exit"`
+	Exited   bool  `json:"exited"`
+
+	Step                        uint64 `json:"step"`
+	StepsSinceLastContextSwitch uint64 `json:"stepsSinceLastContextSwitch"`
+	Wakeup                      uint32 `json:"wakeup"`
+
+	TraverseRight         bool          `json:"traverseRight"`
+	LeftThreadStack       []ThreadState `json:"leftThreadStack"`
+	RightThreadStack      []ThreadState `json:"rightThreadStack"`
+	LeftThreadStackRoots  []common.Hash `json:"leftThreadStackRoots"`
+	RightThreadStackRoots []common.Hash `json:"rightThreadStackRoots"`
+	NextThreadId          uint32        `json:"nextThreadId"`
+
+	// LastHint is optional metadata, and not part of the VM state itself.
+	// It is used to remember the last pre-image hint,
+	// so a VM can start from any state without fetching prior pre-images,
+	// and instead just repeat the last hint on setup,
+	// to make sure pre-image requests can be served.
+	// The first 4 bytes are a uin32 length prefix.
+	// Warning: the hint MAY NOT BE COMPLETE. I.e. this is buffered,
+	// and should only be read when len(LastHint) > 4 && uint32(LastHint[:4]) <= len(LastHint[4:])
+	LastHint hexutil.Bytes `json:"lastHint,omitempty"`
+}
+
+func CreateEmptyMTState() *MTState {
+	initThreadId := uint32(0)
+	initThread := ThreadState{
+		ThreadId: initThreadId,
+		ExitCode: 0,
+		Exited:   false,
+		Cpu: CpuScalars{
+			PC:     0,
+			NextPC: 0,
+			LO:     0,
+			HI:     0,
+		},
+		FutexAddr:        ^uint32(0),
+		FutexVal:         0,
+		FutexTimeoutStep: 0,
+		Registers:        [32]uint32{},
+	}
+	initThreadRoot := computeThreadRoot(EmptyThreadsRoot, &initThread)
+
+	return &MTState{
+		Memory:                NewMemory(),
+		Heap:                  0,
+		ExitCode:              0,
+		Exited:                false,
+		Step:                  0,
+		Wakeup:                ^uint32(0),
+		TraverseRight:         true,
```

Comment:
Can we have threads in the left at init just so `traverseRight` remains default initialized. This is the assumption I made for the contract implementation, though I can change it if you prefer it this way.